### PR TITLE
Updates to the Docker requirements for v2

### PR DESCRIPTION
**Pull request by DavidLiedle**
This PR brings numerous discussions into written form, breaking the requirements into 4 distinct topics:

1. Docker Compose
1. Deployment
1. Data Migration
1. CLI Wrapper

### CODE REVIEW

File: DX/Docker/Compose.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
+# v2: Docker Compose Requirements
+
+## Objective
+To provide a default Docker Compose file or files, ready to use out-of-the-box.
```
Comment (andrerom) [inline DX/Docker/Compose.md]: It's not just docker compose, blue print or whatever we call it also involves docker files (containers), and ways to configure eZ Platform from docker (env variables) which needs changes on eZ Platform itself _(like in 2.0 branch: https://github.com/ezsystems/ezplatform/pull/167/files#diff-bf0e70bcef1a5d5b2f87289220a51108R10)_

File: DX/Docker/Compose.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+1. Place environment in git for collaborative development and versioned with
+  code
+1. Allow new members to team to get started with environment with as few
+  commands as possible
+
+### Preferred
+The feature SHOULD:
+1. Be able to reconfigure eZ Platform via environment variables for all possible
+  server configurations, such as:
+  1. Single/cluster setup
```
Comment (andrerom) [inline DX/Docker/Compose.md]: rending gets wrong here

File: DX/Docker/Wrapper.md [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+Deliverable: The solution shall be one _command_, with options, documented with
+instructions, runnable on Unix, Linux, Mac, and Windows development, staging,
+and production machines, enabling a single approach to common tasks regardless
+of environment.
+
+## Wrapper Solution Requirements
+
+### Firm
+The solution MUST:
+1. Run _inside_ of a Docker container, to unify environments across all systems,
```
Comment (rolandbenedetti) [inline DX/Docker/Wrapper.md]: It seems we have here an issue. On one end, we have the opportunity to build a PHP tool that would deliver most of what we try to achieve, but add the requirement of having PHP running on the command line of the developer environment. On the other, we can still look for a full Docker based solution, but it seems we don't really know how to get there, nor what would be the effort or the technical solution. Pragmatically, I'd like to go for the first route, to quickly enable developers to deploy on their machines. I'd like to revise this requirement expect if we can show a fully Docker based solution that is requiring same level of effort/risk. I have looked at other applications available on Docker hub (such as Drupal or Wordpress) but each time, for anything complex, they end up asking the user to manually do Docker Compose, so I'm not sure about the value.

My vote: be pragmatic and go for a 1st step which require Docker and PHP on CLI.
Insights here highly appreciated.

**Comment by Plopix** [comment]
Hello here,

I think there is no debate on the PHP client approach for the wrapper. PHP is the eZ language of the eZ Community. It will bring more power than bash especially with the Symfony Console and Process components.(YAML component to manage the docker-compose.yml etc.)

Of course, this PHP client MUST not require non-default php-extension (like intl or others). I would say php5.6.x+.

I have concerned about:
> Run inside of a Docker container, to unify environments across all systems, and thus equally usable on Unix, Linux, Windows, and macOS machines.

Did you check the feasibility of this? How I see the way to do it would be to mount the Docker socket inside a PHP container to bring PHP on a system that does not have it. And then inside this container, we would run the PHP Client that would create the stack of the eZ Platform (running docker-compose finally).
I am not sure that this "solution" is correct. I think it is not supposed to work like that I don't this the value.

Something like that would be required:
`docker run -t --rm -v /var/run/docker.sock:/var/run/docker.sock  -w /code -v $(pwd):/code php:5.6-alpine php SOMETING`

The only reason to add this complexity (or this generic way to run the PHP Client) would be because it is hard to install PHP.
=> Which is not to my opinion.

Then comes the Windows case.
- First Docker Toolbox (that installs Virtualbox etc.) is required before a specific build of Windows 10.
- There is not TTY on Windows 

Then the Docker in Docker could be the only solution (unless we find a way to manage interactivity in PHP). For command in the docker stack that will require input.

Maybe the Windows use case must be addressed differently?

File: DX/Docker/Compose.md [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+1. Be able to reconfigure eZ Platform via environment variables for all possible
+  server configurations, such as:
+    1. Single/cluster setup
+    1. Cache using fs/memcached/redis
+    1. Database using mysql/postgres
+    1. Storage using fs/nfs/s3/..
+    1. Search engine using db/solr/es
+    1. Http cache using Symfony or Varnish
+    1. Session stored on disk or in shared backend like memcached/redis/db
+1. Be able to initiate with a database and binary dump
```
Comment (xrow) [inline DX/Docker/Compose.md:30]: https://image.slidesharecdn.com/bjorndiedingrunningezplatformonkubernetes-161011145757/95/running-ez-platform-on-kubernetes-presented-by-bjrn-dieding-at-ez-conference-2016-19-638.jpg?cb=1476197957

This is how we do it in Kubernetes. As the next step we created a Autodiscover Bundle.

File: DX/Docker/README.md [hunk 5]
```diff
@@ -1,91 +1,88 @@
...
+clear focus: a Developer's computer should not consider any requirements, nor
+should staging or production. Encapsulating functionality _within_ containers is
+the key to achieving this goal.
+
+Because containers may be built from images that are ready to run
+out-of-the-box, composed as a pre-configured group of containers, or set up to
+cluster and scale, there are numerous entry points into implementation of these
+options. For eZ Platform, we will begin with a focus on the following:
+
+1. Development / Staging / Production containers, composed by eZ and editable by
```
Comment (xrow) [inline DX/Docker/README.md]: I would say containers as of today are useless for development. If one would show me how I would compare towards vagrant.

File: DX/Docker/Migration.md [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+### Firm
+The feature MUST:
+1. Provide means of dumping/updating my database and files to create a fully
+  working environment.
+1. Run within Docker containers
+
+### Preferred
+The feature SHOULD:
+1. Be able to do this operation from dev to prod, and from prod back to dev
```
Comment (xrow) [inline DX/Docker/Migration.md:26]: Highly desired a matter of the webserver and or init system being able to forward env to the app.

File: DX/Docker/Migration.md [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+For Developers to have a convenient solution for data migration, orchestrated in
+Docker, allowing for zero-configuration actions, sharable with teams.
+
+### Use Cases
+As a Developer, I want...
+1. ...to share database schema changes with my team
+1. ...to migrate production from the old schema to the new one
+1. ...to enjoy identical environments between developers, staging, and production
+
```
Comment (xrow) [inline DX/Docker/Migration.md]: I want to shrink my 50 GB DB

File: DX/Docker/Wrapper.md [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+1. Run _only_ when non-destructive tasks are verified by safety checks in the
+  code, or destructive tasks are confirmed by the user on the CLI with a YES/no
+1. Implement _best practices_ as generally accepted and recommended by the
+  leaders in the Docker community and prevailing usage patterns in the industry
+1. Respect the other requirements set forth in this section (e.g. requirements
+  given in Compose.md, Deployment.md, Migration.md, and the Use Cases in
+  README.md).
+1. Be modular, following the UNIX approach to piping collections of commands, in
+  which each command _does one thing_.
+1. Keep and expose logs to the developer
```
Comment (xrow) [inline DX/Docker/Wrapper.md]: Send the logs to a log service.

**Comment by DavidLiedle** [comment]
New PR from @rolandbenedetti has been merged in, revising the Wrapper requirements. I will make additional updates from the feedback requests from @andrerom and @xrow

File: DX/Docker/Migration.md (hunk 7 above)
Comment (DavidLiedle) [reply DX/Docker/Migration.md]: Adding `...to optimize the size of my DB`

File: DX/Docker/Migration.md (hunk 6 above)
Comment (DavidLiedle) [reply DX/Docker/Migration.md:26]: Adding `Be able to forward ENV variables to the desired app`

File: DX/Docker/Compose.md (hunk 1 above)
Comment (DavidLiedle) [reply DX/Docker/Compose.md]: Updating per feedback, adding mention of Dockerfiles

**Comment by rolandbenedetti** [comment]
This looks good to me.
+1

**Review by rolandbenedetti — APPROVED** [review]

**Pull request merged**